### PR TITLE
feat(maker-core): 智能通讯录两态 system prompt 段(开=使用规范,关=可选功能告示)

### PR DESCRIPTION
**Pull request by zqchris**
### 摘要

智能通讯录此前**零 system prompt 注入**(引导只有工具描述里一句话),agent 经常想不起来用;对比 memory 有 `MAKER_MEMORY_RULES` 注入段,这是两者采用率差距的主要原因。本 PR 补上通讯录的 prompt 段,并做成**两态**:

- **开关开**(~200 token):使用规范——遇人先 `contacts_resolve`、值得记录的信息落档、低置信写 `status:"pending"`、细则服从工具层 `rules` 字段(不重复,只负责"让 agent 想起来用")。
- **开关关**(~150 token):可选功能告示——让 agent 知道功能存在,在三类场景(对话中出现联系人信息更新 / 检索到含联系人信息的内容 / 用户点名找人或搜某人邮件)下可提醒用户一句"可在 设置→个性化→智能通讯录 开启";**每会话至多提一次,被拒不再提并记入 memory**。

接线方式:`BaseAgentDeps` 新增可选 `isContactsEnabled`(host 注入,desktop = `readContactsSettings().enabled`,与 MCP provider 的 `contacts.isEnabled` 同一读数);**缺省时两段都不注入,行为与改造前逐字节一致**。claude-code(本地 sdkQuery + remote cc-manager 两处拼接)与 codex(developerInstructions)同位注入,紧跟 makerMemoryRules 段。

意图契约:目标 = 让所有会话(不改任何渠道入口)天然获得通讯录感知;非目标 = 不改变开关语义、不新增工具、不做自动采集管道。

**system prompt 改动门禁(§4)**:本改动由仓库维护者 @zqchris 在会话中明确提出并逐字审定两态文案后实施("你也可以直接把这个'智能通讯录'的功能写在 system prompt 里面,作为一个可选功能"+ 三类触发场景由维护者本人给出),system prompt 改动已确认。

### 变更类型

- [x] `feat` 新功能
- [ ] `fix` 缺陷修复
- [ ] `refactor` / `perf` 重构或性能优化
- [ ] `docs` / `test` / `chore` 文档、测试或工程维护
- [ ] 其他：

### 范围

- 关联 Issue / 需求：智能通讯录采用性系列第 2 片(第 1 片 #1066,互相独立可合)
- 本 PR 包含：contacts 两态 prompt 模块、BaseAgentDeps 可选字段、两 agent 注入点、desktop host 接线、规则文档拼接顺序同步、内容契约单测
- 明确不包含：任何渠道级 prompt 定制(Telegram bot 等另行处理)、自动采集/定时任务、工具面变化
- 用户可见变化：开启通讯录的用户,agent 会更主动地反查/沉淀联系人;未开启的用户,agent 在相关场景会提一次可开启(每会话至多一次)
- 是否存在 breaking change：无(可选 deps 字段,缺省即旧行为)

### UI 变化

不涉及。

### 自动验证

```text
pnpm test:unit(清除 CLAUDE* 环境变量)
结果:54 PASS / 0 FAIL(含新增 contacts/system-prompt 内容契约测试 3 例)

pnpm --filter desktop run typecheck
结果:通过

pnpm --filter @cindy/maker-core exec vitest run
结果:56 文件 990 用例全过

maker-core tsc --noEmit:38 处既有错误(全部位于 codex/index.test.ts,纯净 main 上
同样 38 处,本 PR 零新增;对比命令:两侧 `npx tsc --noEmit | grep -c "error TS"`)
```

### 核心指标评估(maker-core-and-agent-behavior.md §3.4)

- **可能影响的指标**:缓存率(§3.1,prompt 组装路径)。
- **评估方法与结论**:`isContactsEnabled` 在 startSession 顶层求值一次存入 `const`,claude 两处拼接与 codex 三处 developerInstructions 使用点均复用同一快照(rewind/fork 重建走同一闭包变量),会话内前缀逐字节恒定——与 makerMemoryRules 完全同构,新段为静态文案、无任何 per-turn 易变内容。拼接顺序变化只影响跨版本(部署后首个会话),与既有任何 prompt 文案改动同性质。热路径(§3.2)零触及:求值点在 session 启动,不在 event loop / translator。
- **未执行的实测**:未做真实会话的缓存命中率前后对比(需 live API 会话);风险面已被"同构于 makerMemoryRules + 单次求值 + const 快照"的静态论证覆盖,如需实测可在合并后用 usage-tracker 的 per-turn 命中率抽查。

### 手工验证

不涉及 UI。行为验证建议:开/关通讯录开关各起一个新会话,`/context` 或调试日志确认 system 段分别含 "Smart Contacts" 使用规范/告示段;关态会话中让 agent 查某个联系人邮件,应出现一次开启提醒且同会话不重复。

### 未执行的验证

live 会话实测(开/关两态的实际提醒行为、缓存命中率对比)未做——需要真实模型会话;验证步骤已在上方给出。

### 风险分类

- [ ] 无已知风险
- [x] system prompt
- [ ] 其他：

### 影响与回滚

- 影响范围:所有 desktop 会话的 system 段末尾多一小段静态文案(开态 ~200 / 关态 ~150 token);host 未接线的运行时(如 mobile 直连路径)零变化。
- 回滚方式:revert 单 commit;或 host 侧摘掉 isContactsEnabled 注入即恢复零注入。

### CODE REVIEW

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 4 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts / packages/maker-core/src/agents/claude-code/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 6 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts / packages/maker-core/src/agents/claude-code/index.ts / packages/maker-core/src/contacts/system-prompt-enabled.md),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts [hunk 1]
```diff
@@ -3079,15 +3093,17 @@ export class CodexAgent extends BaseAgent {
...
+      //   [4] contactsRules                    — 智能通讯录两态段 (条件式: host 注入
+      //                                          isContactsEnabled 才有)
+      //   [5] runtimeConfig.systemPrompt       — host runtime (host 维护的 .md)
+      //   [6] makerMemoryIndex                 — 当前 workdir MEMORY.md 内容 (条件式,
       //                                          紧邻 userPrompt 高优先级, 启动时快照)
-      //   [6] opts.userPrompt                  — per-call 用户级 (renderer 本地 storage,
+      //   [7] opts.userPrompt                  — per-call 用户级 (renderer 本地 storage,
       //                                          每次 startSession 透传, 优先级最高)
-      // 跟 claude-code 六段语义对齐。空段被 .filter 跳过,
+      // 跟 claude-code 七段语义对齐。空段被 .filter 跳过,
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts]: 已修(0bcc8f00):claude 头注改为「七段拼接(含 SDK 内嵌 preset)」,codex 尾注标明编号从 [2] 起、共六段且 claude 的 [1] 为 preset,两侧计数口径统一。

File: packages/maker-core/src/agents/claude-code/index.ts [hunk 2]
```diff
@@ -1921,6 +1934,7 @@ export class ClaudeCodeAgent extends BaseAgent {
             const appendText = [
               MAKER_SYSTEM_PROMPT_APPEND,
               makerMemoryRules,
+              contactsRules,
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/claude-code/index.ts:1941]: 已修(0bcc8f00):contactsRules 求值移进 buildQuery,opts.remoteHostId 存在时一律置空——remote 会话两态段都不注入,不再指挥模型调不可达工具。

File: packages/maker-core/src/agents/codex/index.ts [hunk 3]
```diff
@@ -2958,8 +2964,16 @@ export class CodexAgent extends BaseAgent {
...
+    // 智能通讯录两态段: session 启动求值一次, host 未注入 isContactsEnabled 则缺省。
+    // 语义与 claude-code 端一致(contacts/system-prompt.ts)。
+    const contactsRules = this.deps.isContactsEnabled
+      ? this.deps.isContactsEnabled()
+        ? CONTACTS_RULES_ENABLED
+        : CONTACTS_RULES_DISABLED
+      : '';
     const developerInstructions = buildCodexDeveloperInstructions({
       makerMemoryRules,
+      contactsRules,
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts:2983]: 经核实不在本 PR 修:该 marker 的单布尔语义对全部 developerInstructions 段一致——memory 规则或 host prompt 演进同样不会回填已有 thread,本段没有引入新的不一致类别。本地开关切换的一致性已有代码级兜底:MCP flags 冻结于 spawn 配置,contacts-ipc 在开关变更时失效 app-server,新 thread 的工具面与本段按新状态一起重建(0bcc8f00 注释已写明)。把 marker 版本化是跨段的宿主机制升级,应独立评估,不随本 PR 夹带。

File: packages/maker-core/src/agents/claude-code/index.ts [hunk 4]
```diff
@@ -956,6 +960,15 @@ export class ClaudeCodeAgent extends BaseAgent {
...
     }
 
+    // ── 智能通讯录 prompt 段: 开关在 session 启动时求值一次(两态), host 未注入
+    // isContactsEnabled 则整段缺省 — 语义与 makerMemoryRules 同(会话内恒定,
+    // 不破坏前缀缓存; 详见 contacts/system-prompt.ts)。
+    const contactsRules = this.deps.isContactsEnabled
+      ? this.deps.isContactsEnabled()
+        ? CONTACTS_RULES_ENABLED
+        : CONTACTS_RULES_DISABLED
+      : '';
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/claude-code/index.ts]: 已修(0bcc8f00):contactsRules 与 buildMcpServers 同点求值(都在 buildQuery 内读当前设置),rewind/fresh 重建后 prompt 与工具注册不再分叉。

File: packages/maker-core/src/contacts/system-prompt-enabled.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Smart Contacts
+
+The user has Smart Contacts enabled: a local, cross-session people & organization directory served by the `cindy_contacts` MCP server (entry tools: `list_tools` / `call_tool`). Data stays on this device only.
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/contacts/system-prompt-enabled.md]: 已修(0bcc8f00):改为存储与模型处理分开表述——Profile data is stored on this device only; entries you retrieve become part of the conversation context processed by the user's configured model。与 #1066 及 disabled.md 的同款修正对齐。

File: packages/maker-core/src/agents/codex/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts:2983]: 已修(0bcc8f00):codex 侧 opts.remoteHostId 存在时置空,remote 会话不注入两态段。

[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts:2983]: 已修(0bcc8f00):codex 侧 opts.remoteHostId 存在时置空,remote 会话不注入两态段。

File: apps/desktop/src/main/maker-host/index.ts [hunk 6]
```diff
@@ -656,6 +657,9 @@ export function getMaker(): Maker {
       resolveCcDebugFile: resolveSessionCcDebugFile,
       mcpProviders: claudeMcpProviders,
       makerMemory: makerMemoryManager,
+      // 智能通讯录两态 prompt 段的开关快照来源(session start 求值一次), 与
+      // mcp-providers.ts 的 contacts.isEnabled 同一读数, 保证工具面与 prompt 不分叉。
+      isContactsEnabled: () => readContactsSettings().enabled,
```
[the remark this answers is not in the document]
Comment (zqchris) [reply apps/desktop/src/main/maker-host/index.ts]: 已修(00e35b3f):布尔升级为 getContactsPromptState 三态,host 判定链与 provider 包装一致(PluginRegistry 工作区/用户覆盖 → 全局开关);claude 侧 enabled 还与本次 build 实际注册的 server 集合取交。工作区禁用而全局开着 → unavailable,不注入任何段。

File: packages/maker-core/src/agents/codex/index.ts [hunk 7]
```diff
@@ -2958,8 +2964,21 @@ export class CodexAgent extends BaseAgent {
...
+    // remote 会话不注入: 远端 spawn 不带本地 MCP flags、托管 allowlist 只有
+    // collaboration/memory, cindy_contacts 不可达, 注入只会引导模型调不存在的工具。
+    // 本地会话中途切开关的一致性由 host 侧 app-server 失效机制兜底(contacts-ipc):
+    // MCP flags 冻结在 spawn 配置, 开关变更 → 失效 app-server → 新 thread 重建时
+    // 本段与工具面一起按新状态生成。
+    const contactsRules =
+      opts.remoteHostId || !this.deps.isContactsEnabled
+        ? ''
+        : this.deps.isContactsEnabled()
+          ? CONTACTS_RULES_ENABLED
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts]: 已修(00e35b3f):host 在 prepareCodexExtraSpawnConfig 内记录本次 spawn 实际应用的开关快照;live=开而 applied=关(失效失败的 stale 窗口)时返回 unavailable,enabled 段被抑制,直到 app-server 重建成功、快照跟上。bridge 整体构建失败时快照记 false 同样抑制。

File: apps/desktop/src/main/maker-host/index.ts [hunk 8]
```diff
@@ -935,10 +963,15 @@ export function getMaker(): Maker {
           });
           mcpExtraArgs = cfg.extraArgs;
           mcpExtraEnv = cfg.extraEnv;
+          // 本次 spawn 配置实际应用的通讯录开关快照 —— getContactsPromptState 用它
+          // 判定「设置已改但 app-server 失效失败」的 stale 窗口(此时不注入 enabled 段)。
+          codexAppliedContactsEnabled = readContactsSettings().enabled;
```
[the remark this answers is not in the document]
Comment (zqchris) [reply apps/desktop/src/main/maker-host/index.ts]: 已修(6a10f494):快照改为 cfg.bridgeServerNames.includes('cindy_contacts')——从返回的 applied 配置本体推导,缓存命中(pre-toggle 配置)时快照如实反映旧工具面,不再被旁路 live 读数覆盖。指认准确,谢谢。

File: packages/maker-core/src/agents/claude-code/index.ts [hunk 9]
```diff
@@ -2215,21 +2233,24 @@ export class ClaudeCodeAgent extends BaseAgent {
...
+          //   [6] makerMemoryIndex           — 当前 workdir MEMORY.md 内容 (条件式, 紧邻 userPrompt
           //                                    高优先级, 启动时快照 — 跟 userPrompt 同语义)
-          //   [6] opts.userPrompt            — per-call 用户级 (renderer 本地 storage,
+          //   [7] opts.userPrompt            — per-call 用户级 (renderer 本地 storage,
           //                                    每次 startSession 透传, 优先级最高)
           // 空段被 .filter 跳过 (.md 文件为空 / userPrompt 为空 = 不 append).
           systemPrompt: (() => {
             const appendText = [
               MAKER_SYSTEM_PROMPT_APPEND,
               makerMemoryRules,
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/claude-code/index.ts:2252]: 不在本轮跑:per-turn 缓存命中率实测需要带模型凭证的交互式会话,无法在自动化 review 环境完成。已履行部分:(a) 注入机制与位置与已上线的 makerMemoryRules 完全同构(同一 append 列表、同为条件式静态段);(b) 段内容为 per-build 常量、无任何 per-turn 易变值,单次 build 内前缀逐字节恒定由构造保证;(c) PR 验证节如实列明该项为未执行+合并后可用 usage-tracker per-turn 命中率抽查的步骤,未虚报。本 system prompt 改动由仓库维护者 @zqchris 发起并确认,§3.4 的运行时实测口径是否以上述静态论证+事后抽查替代,由维护者裁量——如需合并前实测,请在此 thread 留言,我们补跑后再更新。

File: packages/maker-core/src/agents/claude-code/index.ts [hunk 10]
```diff
@@ -1814,6 +1818,19 @@ export class ClaudeCodeAgent extends BaseAgent {
...
       const mcpServers = buildMcpServers();
+      // ── 智能通讯录两态段: 紧跟 buildMcpServers 求值, prompt 状态与本次 build
+      // 实际生效的工具面对齐(rewind/fresh 重建同步跟随), 三种去向:
+      //   host 有效状态 enabled 且 cindy_contacts 真的注册了 → 使用规范段;
+      //   disabled(功能未开) → 可选功能告示段(邀请开启);
+      //   其余(unavailable/工作区覆盖禁用/注册被跳过/remote/未接线) → 不注入 —
+      //   既不指挥模型调不可达工具, 也不邀请用户去开一个已开着的开关。
+      const contactsRules = (() => {
+        if (opts.remoteHostId) return '';
+        const state = this.deps.getContactsPromptState?.({ workingDir: opts.workingDir });
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/claude-code/index.ts:1829]: 经核实不改:这条与第 1 轮 greptile 的已确认 P1(Rewind 后开关状态分叉)方向正好相反,两者不可同时满足,需要一次性的设计裁定,裁定如下并已在代码中实现——

**不变量 = 「prompt 状态永远与同一次 build 实际生效的工具面一致」,而非「整个逻辑会话冻结」。**

理由:
1. buildMcpServers() 在每次 rewind/fresh 重建时 live 重估 provider.isEnabled() 是本 PR 之前就存在的基线行为(对 lsp、自定义 MCP 等所有带门禁的 provider 同样成立),不在本 PR 范围内改变。在这个基线下,只冻结 contactsRules 会精确复现第 1 轮已确认的分叉:toggle+rewind 后 prompt 指挥模型用已移除的工具(或反向)。工具面与 prompt 的一致性是正确性问题,优先于缓存。
2. 缓存代价有界:前缀变化只发生在「用户真的切了开关 + 又触发了 rewind」的罕见窗口,一次性损失一个 resume 点的缓存;正常会话(不切开关)下每次 rebuild 求值结果恒定,前缀逐字节稳定。
3. 「settings handler 的 next-session-start 生命周期」在基线里本就被 rewind 重建打破(工具面会变),prompt 跟随工具面恰恰是与既有生命周期语义一致的唯一选择。

如认为应改为全会话冻结,需要连同 buildMcpServers 的重建语义一起改,那是独立于本 PR 的基线变更,应另行评估。

**Review by MagicLizi — APPROVED** [review]
Review passed: two-state contacts prompt follows established patterns, handles enabled/disabled/unavailable edge cases correctly. Good test coverage.

**Comment by MagicLizi** [comment]
通讯录提示的三态模型(开/关/不可用)把各种部署形态的边界都考虑到了,prompt 与 MCP 的同步也干净利落。

**Comment by zqchris** [comment]
## Live 两态实测结果(隔离 dev 沙盒,head 6a10f49)

按 §3.4 补跑了真实会话验证。环境:`--isolated` 沙盒(独立 userData)、local 模式(跳过登录)、Claude Code / Opus 4.8,通过预写 `contacts-settings.json` 切两态、CDP 驱动会话,证据取自 per-session ndjson 日志与本地用量 DB。

**开态(enabled:true,会话 399e2620)**
- 让模型逐字引用 system prompt 中 Smart Contacts 相关段 → **逐字引用了 CONTACTS_RULES_ENABLED 全文**(resolve-first / pending 语义 / rules 服从,一字不差)。
- 缓存:turn1 `cache_create=56190`;turn2 `cache_read=56190`(**上一轮前缀 100% 命中**),增量 create 仅 429。前缀会话内逐字节稳定。

**关态(enabled:false,会话 3d6aad06)**
- maker 侧探针(临时 log,已移除)证实注入判定:`hasDep:true, state:disabled, remote:false, bytes:969`。
- 模型逐字引用了 **CONTACTS_RULES_DISABLED 全文**(含设置路径与「每会话至多一次」约束)。
- 缓存:同 boot 内跨会话前缀共享正常(turn1 `cache_read=46276` + `create=8729`,turn hitRate 84.1%),无异常膨胀。

结论:两态注入内容正确、开关切换经重启会话后生效、缓存命中行为健康。此前 thread 中declared 的「未执行 live 实测」项已补齐。

(实测过程记录:早期两次误报 NO_CONTACTS_SECTION 系测试脚手架污染——把消息发进了恢复的旧会话;带探针的干净复跑为准。)

File: packages/maker-core/src/agents/claude-code/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/claude-code/index.ts:2252]: 已按承诺补跑 live 实测,结果见 PR 评论(两态注入逐字命中 + turn2 前缀 100% 缓存命中):https://github.com/makecindy/cindy/pull/1069#issuecomment-5131457866
